arangod/Aql/RemoteExecutor.cpp

```diff
@@ -350,171 +341,106 @@ std::pair<ExecutionState, Result> ExecutionBlockImpl<RemoteExecutor>::shutdown(i
-      std::shared_ptr<VPackBuilder> builder =
-          VPackParser::fromJson(responseBodyBuf.c_str(), responseBodyBuf.length());
-      VPackSlice slice = builder->slice();
-
-      if (!slice.hasKey(StaticStrings::Error) ||
-          slice.get(StaticStrings::Error).getBoolean()) {
-        errorNum = TRI_ERROR_INTERNAL;
-      }
-
-      if (slice.isObject()) {
-        VPackSlice v = slice.get(StaticStrings::ErrorNum);
-        if (v.isNumber()) {
-          if (v.getNumericValue<int>() != TRI_ERROR_NO_ERROR) {
-            /* if we've got an error num, error has to be true. */
-            TRI_ASSERT(errorNum == TRI_ERROR_INTERNAL);
-            errorNum = v.getNumericValue<int>();
-          }
-        }
+  std::string url = std::string("/_db/") +
+                    arangodb::basics::StringUtils::urlEncode(
+                        _engine->getQuery()->trx()->vocbase().name()) +
+                    urlPart + _queryId;
 
-        v = slice.get(StaticStrings::ErrorMessage);
-        if (v.isString()) {
-          errorMessage += v.copyString();
-        } else {
-          errorMessage += std::string("(no valid error in response)");
-        }
-      }
-    }
-    // In this case a proper HTTP error was reported by the DBserver,
-    if (errorNum > 0 && !errorMessage.empty()) {
-      return {errorNum, errorMessage};
-    }
+  arangodb::network::EndpointSpec spec;
+  int res = network::resolveDestination(_server, spec);
+  if (res != TRI_ERROR_NO_ERROR) {  // FIXME return an error  ?!
+    return Result(res);
+  }
+  TRI_ASSERT(!spec.endpoint.empty());
 
-    // default error
-    return {TRI_ERROR_CLUSTER_AQL_COMMUNICATION};
+  auto req = fuerte::createRequest(type, url, {}, std::move(body));
+  // Later, we probably want to set these sensibly:
+  req->timeout(std::chrono::seconds(kDefaultTimeOutSecs));
+  if (!_ownName.empty()) {
+    req->header.addMeta("Shard-Id", _ownName);
   }
 
-  TRI_ASSERT(res->status == CL_COMM_SENT);
+  network::ConnectionPool::Ref ref = pool->leaseConnection(spec.endpoint);
+
+  unsigned ticket = _lastTicket.fetch_add(1) + 1;
+  std::shared_ptr<fuerte::Connection> conn = ref.connection();
+  conn->sendRequest(std::move(req),
+                    [=, ref(std::move(ref))](fuerte::Error err,
+                                             std::unique_ptr<fuerte::Request>,
+                                             std::unique_ptr<fuerte::Response> res) {
+                      _query.sharedState()->execute([&] {  // notifies outside
+                        if (_lastTicket.load() == ticket) {
+                          if (err != fuerte::Error::NoError || res->statusCode() >= 400) {
+                            _lastError = handleErrorResponse(spec, err, res.get());
+                          } else {
+                            _lastResponse = std::move(res);
+                          }
+                        }
+                      });
+                    });
```

Comment:
@graetzer You removed the mutex protecting `_lastError`, `_lastResponse` and the ticket id. It is not obvious to me that this is safe. Please add an argument why this is so. Also, I think @mchacki should have a look at that, because he thought a lot about possible races in the RemoteExecutor.

Reply:
@graetzer @goedderz yes we had this race-conditions in the past.
I think even with this code we might get into the race:
Assume we send a getSomeRequest (that for some reason takes long) (A)
We then trigger the shutdown because of Timeout. (B)

Is it possible that this now generates the following sequence:
(A) callback comes in, checks ticketID => success
(B) takes over, and clears, _lastError/_lastResponse, generates _lastTicket
(A) takes over, set's whatever it has. Awakens the query
(B) returns the stack, is rescheduled
(B) wakes up, and mistakes whatever (A) has written as response to (B)?

If this can happen we get into issues later as the format is not correct. and we might end-up with false-positives / false negatives